### PR TITLE
Make ClusterQueue.spec.queuingStrategy immutable

### PR DESCRIPTION
**Pull request by cmssczy**
Signed-off-by:  cmssczy <smartczy@outlook.com>
#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes # https://github.com/kubernetes-sigs/kueue/issues/434

### CODE REVIEW

File: apis/kueue/v1alpha2/clusterqueue_types.go [hunk 1]
```diff
@@ -134,6 +137,7 @@ type ClusterQueueSpec struct {
 	// admitting newer workloads that fit existing quota.
 	//
 	// +kubebuilder:default=BestEffortFIFO
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
```
Comment (kerthcet) [inline apis/kueue/v1alpha2/clusterqueue_types.go:140]: I think it should be `QueueingStrategy is immutable`

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 2]
```diff
@@ -108,6 +111,14 @@ func ValidateClusterQueue(cq *kueue.ClusterQueue) field.ErrorList {
 	return allErrs
 }
 
+// We need to do this before Kubernetes 1.25.
+func ValidateClusterQueueUpdate(newObj, oldObj *kueue.ClusterQueue) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.QueueingStrategy, oldObj.Spec.QueueingStrategy, field.NewPath("spec", "queueingStrategy"))...)
+	allErrs = append(allErrs, ValidateClusterQueue(newObj)...)
```
Comment (kerthcet) [inline apis/kueue/webhooks/clusterqueue_webhook.go:118]: Let's validate clusterQueue first.

File: apis/kueue/v1alpha2/clusterqueue_types.go [hunk 3]
```diff
@@ -124,6 +124,9 @@ type ClusterQueueSpec struct {
 
 	// QueueingStrategy indicates the queueing strategy of the workloads
 	// across the queues in this ClusterQueue. This field is immutable.
+	// Since Kubernetes 1.25, we can use validation rules to implement
```
Comment (kerthcet) [inline apis/kueue/v1alpha2/clusterqueue_types.go:127]: s/validation rules/CEL validation rules/

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 4]
```diff
@@ -108,6 +111,14 @@ func ValidateClusterQueue(cq *kueue.ClusterQueue) field.ErrorList {
 	return allErrs
 }
 
+// We need to do this before Kubernetes 1.25.
+func ValidateClusterQueueUpdate(newObj, oldObj *kueue.ClusterQueue) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newObj.Spec.QueueingStrategy, oldObj.Spec.QueueingStrategy, field.NewPath("spec", "queueingStrategy"))...)
```
Comment (kerthcet) [inline apis/kueue/webhooks/clusterqueue_webhook.go:117]: Remove the comment above and add here:
// We need to validate the spec.queueingStrategy immutable manually before Kubernetes 1.25.

File: apis/kueue/v1alpha2/clusterqueue_types.go [hunk 5]
```diff
@@ -124,6 +124,9 @@ type ClusterQueueSpec struct {
 
 	// QueueingStrategy indicates the queueing strategy of the workloads
 	// across the queues in this ClusterQueue. This field is immutable.
+	// Since Kubernetes 1.25, we can use CEL validation rules to implement
+	// a few common immutability patterns directly in the manifest for a CRD.
+	// ref: https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/
```
Comment (alculquicondor) [inline apis/kueue/v1alpha2/clusterqueue_types.go:129]: This is part of the user documentation, don't add this kind of detail here. Maybe leave it in the webhook code.

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 6]
```diff
@@ -108,6 +111,14 @@ func ValidateClusterQueue(cq *kueue.ClusterQueue) field.ErrorList {
 	return allErrs
 }
 
+// We need to validate the spec.queueingStrategy immutable manually before Kubernetes 1.25.
```
Comment (alculquicondor) [inline apis/kueue/webhooks/clusterqueue_webhook.go:114]: Could you create an issue for this? I think we can remove the webhooks in favor of CEL once 1.24 reaches EoL https://kubernetes.io/releases/

File: apis/kueue/webhooks/clusterqueue_webhook_test.go [hunk 7]
```diff
@@ -236,3 +236,36 @@ func TestValidateClusterQueue(t *testing.T) {
...
+		newClusterQueue *kueue.ClusterQueue
+		oldClusterQueue *kueue.ClusterQueue
+		wantErr         field.ErrorList
+	}{
+		{
+			name:            "queueingStrategy cannot be updated",
+			newClusterQueue: testingutil.MakeClusterQueue("cluster-queue").QueueingStrategy("BestEffortFIFO").Obj(),
+			oldClusterQueue: testingutil.MakeClusterQueue("cluster-queue").QueueingStrategy("StrictFIFO").Obj(),
+			wantErr: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("queueingStrategy"), nil, ""),
```
Comment (alculquicondor) [inline apis/kueue/webhooks/clusterqueue_webhook_test.go:252]: I think this works?
```suggestion
				field.Invalid(field.NewPath("spec", "queueingStrategy"), nil, ""),
```

File: apis/kueue/webhooks/clusterqueue_webhook.go (hunk 6 above)
Comment (cmssczy) [reply apis/kueue/webhooks/clusterqueue_webhook.go:117]: https://github.com/kubernetes-sigs/kueue/issues/463

**Comment by kerthcet** [comment]
/lgtm
Thanks @cmssczy

File: apis/kueue/v1alpha2/clusterqueue_types.go [hunk 8]
```diff
@@ -134,6 +134,7 @@ type ClusterQueueSpec struct {
 	// admitting newer workloads that fit existing quota.
 	//
 	// +kubebuilder:default=BestEffortFIFO
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="QueueingStrategy is immutable"
```
Comment (alculquicondor) [inline apis/kueue/v1alpha2/clusterqueue_types.go:137]: I prefer we don't add the CEL validations while we still have the webhooks.

Unless we want to allow users to install kueue without webhooks. But then, we should have **all** the validations implemented in CEL, not just this one.

**Pull request merged**